### PR TITLE
simulator related docstr refine

### PR DESCRIPTION
**Pull request by chaosddp**

### CODE REVIEW

File: examples/hello_world/citi_bike/hello.py [hunk 1]
```diff
@@ -23,12 +23,12 @@
     while not is_done:
         metrics, decision_evt, is_done = env.step(action)
 
-        if decision_evt is not None: # it will be None at the end
+        if decision_evt is not None:  # it will be None at the end
```
Comment (Jinyu-W) [inline examples/hello_world/citi_bike/hello.py:26]: Move to the line above

File: examples/hello_world/citi_bike/hello.py [hunk 2]
```diff
@@ -41,7 +41,8 @@
 
     matrix_ss = env.snapshot_list["matrices"]
 
-    # since we may have different snapshot resolution, so we should use frame_index to retrieve index in snapshots of current tick
+    # since we may have different snapshot resolution,
```
Comment (Jinyu-W) [inline examples/hello_world/citi_bike/hello.py:44]: Use "Since" instead of “since”. Add a period at the end.

File: maro/backends/frame.pxd [hunk 3]
```diff
@@ -61,26 +70,70 @@ cdef class SnapshotList:
 
 
 cdef class FrameBase:
-    """Wrapper to define Frame as a class to make it simple to write.
+    """Base object used to define frame in backend, any frame that need to be hosted in backend should inherit from this.
+    Usually a frame is composed with serveral nodes (NodeBase), a snapshot list if enabled.
```
Comment (Jinyu-W) [inline maro/backends/frame.pxd:74]: Insert a blank line above.

File: maro/backends/frame.pxd [hunk 4]
```diff
@@ -61,26 +70,70 @@ cdef class SnapshotList:
...
+        my_nodes_list = frame.mynodes
+        your_nodes_list = frame.yournodes
+
+        for mnode in my_node_list:
+            print(mnode.index) # 0 - len(my_node_list)
 
     
     Args:
-        enable_snapshot (bool): if enable snapshot to keep Frame states, default False
+        enable_snapshot (bool): if enable snapshot list to keep frame snapshot at specified point, default False
```
Comment (Jinyu-W) [inline maro/backends/frame.pxd:130]: Add a period (``.``) to the end of each args

File: maro/backends/frame.pxd [hunk 5]
```diff
@@ -61,26 +70,70 @@ cdef class SnapshotList:
...
+    When defining a frame, number of node must be specified, this may not suitable for all the case, such as the node number
+    is dynamic generated after initializing, this is a workaround to fix this that use function wrapper to generate frame 
+    definition at runtime.
 
-                def __init__(self, enable_snapshot:bool=True, snapshot_number: int = 10):
-                    super().__init__(self, enable_snapshot, total_snapshots=snapshot_number)
+
+    .. code-block:: python
+
+        # using function to wrap the frame definition to support dynamic node number
```
Comment (Jinyu-W) [inline maro/backends/frame.pxd:100]: Start with a capital letter, end with a period. Same to others

**Review by Jinyu-W — CHANGES_REQUESTED** [review]

File: examples/hello_world/citi_bike/hello.py [hunk 6]
```diff
@@ -23,12 +23,11 @@
     while not is_done:
         metrics, decision_evt, is_done = env.step(action)
 
-        if decision_evt is not None: # it will be None at the end
+        # it will be None at the end
```
Comment (Jinyu-W) [inline examples/hello_world/citi_bike/hello.py:26]: It .... .(period)

File: examples/hello_world/citi_bike/hello.py [hunk 7]
```diff
@@ -41,7 +40,8 @@
 
     matrix_ss = env.snapshot_list["matrices"]
 
-    # since we may have different snapshot resolution, so we should use frame_index to retrieve index in snapshots of current tick
+    # Since we may have different snapshot resolution,
+    # so we should use frame_index to retrieve index in snapshots of current tick
```
Comment (Jinyu-W) [inline examples/hello_world/citi_bike/hello.py:44]: end with a . period

Comment (Jinyu-W) [reply examples/hello_world/citi_bike/hello.py:44]: Same to other comments

File: maro/backends/frame.pxd [hunk 8]
```diff
@@ -61,26 +70,72 @@ cdef class SnapshotList:
...
+
+        for mnode in my_node_list:
+            print(mnode.index) # 0 - len(my_node_list)
 
     
     Args:
-        enable_snapshot (bool): if enable snapshot to keep Frame states, default False
-        total_snapshots (int): total snapshots number in memory
-        options (dict): additional options, reserved for later using
+        enable_snapshot (bool): if enable snapshot list to keep frame snapshot at specified point, default False.
```
Comment (Jinyu-W) [inline maro/backends/frame.pxd:132]: Start with Capital letter. Same to others.

File: maro/backends/frame.pxd [hunk 9]
```diff
@@ -101,34 +156,126 @@ cdef class FrameBase:
...
+    """Helper class used to define a node in frame, specified node class and number.
     
+
     Usually use with FrameBase, see FrameBase for details.
     
+    
     Args:
-        node_cls(type): class type of node definition inherit from NodeBase
-        number(int): the number of this node in Frame
+        node_cls(type): class type of node definition inherit from NodeBase.
```
Comment (Jinyu-W) [inline maro/backends/frame.pxd:166]: Start with Capital letter. Same to others.

File: maro/backends/frame.pxd [hunk 10]
```diff
@@ -101,34 +156,126 @@ cdef class FrameBase:
...
 
 
 cdef class NodeBase:
-    """Helper to provide easy way to define a node in Frame.
-    
-    Example:
-        .. code-block:: python
+    """Helper class used to define a node with attributes, any node that need to be hosted in backend should 
+    inherit from this class.
+    A node is composed with serveral attributes that defined with NodeAttribute class with data type and slot number.
```
Comment (Jinyu-W) [inline maro/backends/frame.pxd:178]: An blank line is needed above line 178.

File: maro/backends/frame.pxd [hunk 11]
```diff
@@ -101,34 +156,126 @@ cdef class FrameBase:
...
 cdef class NodeBase:
-    """Helper to provide easy way to define a node in Frame.
-    
-    Example:
-        .. code-block:: python
+    """Helper class used to define a node with attributes, any node that need to be hosted in backend should 
+    inherit from this class.
+    A node is composed with serveral attributes that defined with NodeAttribute class with data type and slot number. 
+        
+
```
Comment (Jinyu-W) [inline maro/backends/frame.pxd:180]: To many blank lines here?

File: maro/backends/frame.pxd [hunk 12]
```diff
@@ -101,34 +156,126 @@ cdef class FrameBase:
...
+
+    .. code-block:: python
 
-            # node name in Frame, we use this name to query from snapshot list
+        # Node name in Frame, we use this name to query from snapshot list.
+        @node("my nodes")
+        class MyNode(NodeBase): # node class we used in business engine
+            # Attribute name, and its data type.
+            # int attribute with 1 value
```
Comment (Jinyu-W) [inline maro/backends/frame.pxd:197]: Capital, period ?

File: maro/backends/frame.pxd [hunk 13]
```diff
@@ -101,34 +156,126 @@ cdef class FrameBase:
...
+            my_int_attr = NodeAttribute("i")
+
+            # a fixed size float array
+            my_float_array_attr = NodeAttribute("f", 2)
+
+
+    Each attribute will have a default hook than will trigger an event after the value changed, to recieve this event
+    there should be a specified method that named as '_on_<attribute name>_changed' in node definition class,
```
Comment (Jinyu-W) [inline maro/backends/frame.pxd:205]: use double \` (\`\`key_words\`\`) here?

File: maro/backends/frame.pxd [hunk 14]
```diff
@@ -150,17 +297,19 @@ cdef class NodeBase:
...
     
-    See NodeBase for details.
+
+    Currently we only support these data types: 'i', 'i2', 'i4', 'i8', 'f', 'd'
+
 
     Args:
-        dtype(str): type of this attribute, it support following data types like numpy: 'i', 'i2', 'i4', 'i8', 'f', 'd'
-        slots(int): if this number greater than 1, then it will be treat as an array, this is the array size.
+        dtype(str): type of this attribute, it support following data types.
```
Comment (Jinyu-W) [inline maro/backends/frame.pxd:307]: Capital letter

File: maro/backends/frame.pyx [hunk 15]
```diff
@@ -39,16 +40,16 @@ def node(name: str):
...
 cdef class NodeAttribute:
-    def __cinit__(self, dtype: str, slot_num: int = 1):
-        self.dtype = dtype
-        self.slot_number = slot_num
+    def __init__(self, dtype: str, slot_num: int = 1):
+        self._dtype = dtype
+        self._slot_number = slot_num
 
 
 # TODO: a better way to support multiple value get/set for an attribute with more than one slot
```
Comment (Jinyu-W) [inline maro/backends/frame.pyx:48]: end with period (.)

File: maro/backends/frame.pyx [hunk 16]
```diff
@@ -185,20 +186,20 @@ cdef class NodeBase:
...
         cdef dict __dict__ = object.__getattribute__(self, "__dict__")    
 
         cdef str name
         cdef str node_name
         cdef str cb_name
-        cdef NodeAttributeAccessor attr_acc
+        cdef _NodeAttributeAccessor attr_acc
 
         for name, attr in type(self).__dict__.items():
             # append an attribute access wrapper to current instance
```
Comment (Jinyu-W) [inline maro/backends/frame.pyx:205]: Start with Capital letter, end with period.

File: maro/backends/frame.pyx [hunk 17]
```diff
@@ -268,30 +269,62 @@ cdef class FrameBase:
...
+        """Reset internal states of frame, currently all the attributes will reset to 0.
+        
+        Note:
+            This method will not reset states in snapshot list.
+        """
         self._backend.reset()
 
     cpdef void take_snapshot(self, int tick) except *:
+        """Take snapshot for specified point (tick) for current frame. 
+        This method will copy current frame value into snapshot list for later using.
```
Comment (Jinyu-W) [inline maro/backends/frame.pyx:293]: A blank line is needed above this

File: maro/backends/frame.pyx [hunk 18]
```diff
@@ -268,30 +269,62 @@ cdef class FrameBase:
...
+
+
```
Comment (Jinyu-W) [inline maro/backends/frame.pyx:295]: Too many blank lines

**Review by Jinyu-W — CHANGES_REQUESTED** [review]

File: maro/backends/frame.pyx (hunk 16 above)
Comment (chaosddp) [reply maro/backends/frame.pyx:205]: this is an internal class, here use _ to hide it in api doc.

**Pull request merged**